### PR TITLE
Command pimcore:thumbnails:image / get thumbnail image of remotely stored assets

### PR DESCRIPTION
**Pull request by BlackbitDevs**
When asset files are stored on a cloud hosting provider (in our case Google Cloud Storage), there was an error 
```
[pimcore] ImagickException: insufficient image data in file `/var/www/html/var/tmp/asset-temporary/asset_739_30573635c4f669ded2c13782092cf497__xyz.jpg' @ error/jpeg.c/ReadJPEGImage/1117 in /var/www/html/vendor/pimcore/pimcore/lib/Image/Adapter/Imagick.php:126
```

This happens when the stream cannot be rewinded - but until now the stream was not reopened in this case when the file was not locally, see https://github.com/pimcore/pimcore/blob/234360e575cdc6cdc6d812082b6585055817d882/models/Asset.php#L1322

While I fixed this I stumbled upon https://github.com/pimcore/pimcore/blob/234360e575cdc6cdc6d812082b6585055817d882/lib/Image/Adapter/Imagick.php#L61-L75
As this was too hard-coded to HTTP(S) and AWS S3 in my opinion I changed it in analogy to https://github.com/pimcore/pimcore/blob/234360e575cdc6cdc6d812082b6585055817d882/models/Asset.php#L1553-L1570

And I removed https://github.com/pimcore/pimcore/blob/234360e575cdc6cdc6d812082b6585055817d882/lib/Image/Adapter/Imagick.php#L81 because this temporary file gets already handled by https://github.com/pimcore/pimcore/blob/234360e575cdc6cdc6d812082b6585055817d882/models/Asset.php#L2029-L2033

### CODE REVIEW

File: lib/Templating/Renderer/IncludeRenderer.php [hunk 1]
```diff
@@ -89,7 +89,7 @@ public function render($include, array $params = [], $editmode = false, $cacheEn
         $cacheKey = null;
         $cacheConfig = false;
 
-        if ($cacheEnabled) {
+        if ($cacheEnabled && !$editmode) {
```
Comment (brusch) [inline lib/Templating/Renderer/IncludeRenderer.php:92]: There seems to be something wrong - this changes come from a commit which doesn't belong to this PR. 
@BlackbitNeueMedien can you please check? 
Thanks!

**Review by brusch — CHANGES_REQUESTED** [review]

**Comment by BlackbitNeueMedien** [comment]
Sorry, wrong base branch. Now it is correct, @brusch

File: lib/Templating/Renderer/IncludeRenderer.php (hunk 1 above)
Comment (BlackbitNeueMedien) [reply lib/Templating/Renderer/IncludeRenderer.php]: Sorry, wrong base branch. Now it is correct, @brusch

Comment (brusch) [reply lib/Templating/Renderer/IncludeRenderer.php]: 👍 That looks better, thanks a lot!

**Review by brusch — CHANGES_REQUESTED** [review]
Would it be possible for you to fix those issue in other places as well? 
https://github.com/pimcore/pimcore/blob/9b0a4ebdc831dbd5cc21d314388ee34b8d4f8703/lib/Document/Adapter.php#L33
https://github.com/pimcore/pimcore/blob/d132eda35ef6b5278c2b8e0becbfc8c25e7924da/lib/Video/Adapter/Ffmpeg.php#L82
https://github.com/pimcore/pimcore/blob/05bbe5f784a5b32b9f161a6277282495c238fe0f/models/Asset/Image.php#L151
https://github.com/pimcore/pimcore/blob/947e4b62e48610bb71615e39d937a3e67287258f/lib/Image/Optimizer.php#L101
https://github.com/pimcore/pimcore/blob/e2d8837a46ce0a99711be8cc5927418bb221286b/bundles/AdminBundle/Controller/Admin/AssetController.php#L1213
https://github.com/pimcore/pimcore/blob/4af305c31d20d6f2c1e48056538b8e669f3d9d3f/models/Asset/MetaData/EmbeddedMetaDataTrait.php#L71
https://github.com/pimcore/pimcore/blob/05bbe5f784a5b32b9f161a6277282495c238fe0f/models/Asset/Image.php#L211

File: lib/Image/Adapter/Imagick.php [hunk 2]
```diff
@@ -58,27 +58,24 @@ public function load($imagePath, $options = [])
...
-        if (preg_match('@^https?://|^s3://@', $imagePath)) {
-            $tmpFilename = 'imagick_auto_download_' . md5($imagePath) . '.' . File::getFileExtension($imagePath);
-            $tmpFilePath = PIMCORE_SYSTEM_TEMP_DIRECTORY . '/' . $tmpFilename;
-
-            $this->tmpFiles[] = $tmpFilePath;
-
-            if (preg_match('@^s3://@', $imagePath)) {
-                $imageContent = file_get_contents($imagePath);
+        if (!stream_is_local($imagePath)) {
+            if (isset($options['asset'])) {
```
[the remark this answers is not in the document]
Comment (BlackbitNeueMedien) [reply lib/Image/Adapter/Imagick.php:62]: But for locally stored images we actually do not need a temporary file. Or have I understood you wrong?

[the remark this answers is not in the document]
Comment (brusch) [reply lib/Image/Adapter/Imagick.php:62]: I think you got me wrong :) 
What I meant is that this code does actually the same thing: 
```php
if (!stream_is_local($imagePath)) {
        $tmpFilename = 'imagick_auto_download_'.md5($imagePath).'.'.File::getFileExtension($imagePath);
        $tmpFilePath = PIMCORE_SYSTEM_TEMP_DIRECTORY.'/'.$tmpFilename;

        $this->tmpFiles[] = $tmpFilePath;

        $src = fopen($imagePath, 'rb');
        $dest = fopen($tmpFilePath, 'wb', false, File::getContext());
        stream_copy_to_stream($src, $dest);
        fclose($dest);

        $imagePath = $tmpFilePath;
    }
}
``` 

we don't need this part: 
```php
    if (isset($options['asset'])) {
        // imagick is only able to deal with local files
        // if your're using custom stream wrappers this wouldn't work, so we create a temp. local copy
        $imagePath = $options['asset']->getTemporaryFile();
    } else {
``` 

because if `$imagePath` isn't a local stream we have to create a temp file anyway 😊

[the remark this answers is not in the document]
Comment (BlackbitNeueMedien) [reply lib/Image/Adapter/Imagick.php]: Ok, I removed this condition.

File: lib/Image/Adapter/Imagick.php [hunk 3]
```diff
@@ -58,27 +58,24 @@ public function load($imagePath, $options = [])
...
-            $imagePath = $tmpFilePath;
-        }
+                $this->tmpFiles[] = $tmpFilePath;
 
-        if (!stream_is_local($imagePath) && isset($options['asset'])) {
-            // imagick is only able to deal with local files
-            // if your're using custom stream wrappers this wouldn't work, so we create a temp. local copy
-            $imagePath = $options['asset']->getTemporaryFile();
-            $this->tmpFiles[] = $imagePath;
+                $src = fopen($imagePath, 'rb');
```
[the remark this answers is not in the document]
Comment (BlackbitNeueMedien) [reply lib/Image/Adapter/Imagick.php]: Hmm, how can we care about those temporary files? I do not want to implement `__destruct()` in a trait. And otherwise if we implement the cleanup in a separate method, we always have to think of the cleanup. Perhaps it would be better to use a separate class like 
```php
class TemporaryFileHelper {
  private $tmpFiles = [];
  
  public function __destruct() {
    foreach ($this->tmpFiles as $tempFile) {
            if (file_exists($tempFile)) {
                @unlink($tempFile);
            }
        }
  }

  public function getLocalFile($path) {
    if (!stream_is_local($path)) {
            $tmpFilename = 'tmpfile_'.md5($path).'.'.File::getFileExtension($path);
            $tmpFilePath = PIMCORE_SYSTEM_TEMP_DIRECTORY.'/'.$tmpFilename;

            self::tmpFiles[] = $tmpFilePath;

            $src = fopen($path, 'rb');
            $dest = fopen($tmpFilePath, 'wb', false, File::getContext());
            stream_copy_to_stream($src, $dest);
            fclose($dest);

            return $tmpFilePath;
        }

       return $path;
  }
}

File: lib/Image/Adapter/Imagick.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (BlackbitNeueMedien) [reply lib/Image/Adapter/Imagick.php]: Ok, implemented for all files you have mentioned except https://github.com/pimcore/pimcore/blob/f093a1893b83495bc864ea5d89d25a42b15e8a72/lib/Image/Optimizer.php#L101
There I did not understand where a local file is necessary.

**Comment by brusch** [comment]
@BlackbitNeueMedien thanks a lot for your changes!
I think  `Asset::getTemporaryFile()` should also utilize the new trait. 
Regarding `__destruct()` ... maybe a temp-file cleanup service is the solution, that cleans up all registered tempfiles on shutdown? wdyt?

**Comment by BlackbitNeueMedien** [comment]
Also changed `Asset::getTemporaryFile()`.

The cleanup of the temporary files gets handled via 
```php
register_shutdown_function(static function() use ($tmpFilePath) {
    @unlink($tmpFilePath);
});
```
within the trait, so we do not need to implement this cleanup in a `__destruct()` method.

**Comment by BlackbitNeueMedien** [comment]
@brusch Do you have an idea why the test no fails?
```Fatal error: Class Symfony\Component\DependencyInjection\Exception\ExceptionInterface cannot implement previously implemented interface Throwable in /home/runner/work/pimcore/pimcore/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Exception/ExceptionInterface.php on line 22 ```
This has nothing to do with the changes of this PR.

**Comment by jdreesen** [comment]
It was a bug in [PSR-11 (Container Interface)](https://github.com/php-fig/container) v1.1, which was released yesterday. 
[It has been fixed in v1.1.1](https://github.com/php-fig/container/releases/tag/1.1.1) an nothing to do with this PR.

**Comment by BlackbitNeueMedien** [comment]
Thx @jdreesen - now tests are green.

File: lib/Image/Adapter/Imagick.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (brusch) [reply lib/Image/Adapter/Imagick.php]: You're right, https://github.com/pimcore/pimcore/blob/f093a1893b83495bc864ea5d89d25a42b15e8a72/lib/Image/Optimizer.php#L101 is fine as it is.

**Comment by brusch** [comment]
LGTM, I've found some more potential for optimization, but I'll do that in a follow up PR.
Thanks a lot!

**Comment by supryantowp** [comment]
> 

May I ask. i'm having the same problem now i'm using pimcore version ~6.5.0. how can i fix this problem

**Comment by supryantowp** [comment]
> It was a bug in [PSR-11 (Container Interface)](https://github.com/php-fig/container) v1.1, which was released yesterday. [It has been fixed in v1.1.1](https://github.com/php-fig/container/releases/tag/1.1.1) an nothing to do with this PR.

May I ask. i'm having the same problem now i'm using pimcore version ~6.5.0. how can i fix this problem
